### PR TITLE
Rate limiting tab

### LINKED ISSUE #981: Separate Rate limit settings into its own tab
Self explanatory.

### PR DESCRIPTION
**Pull request by SleeplessOne1917**
This closes #981. This is what the rate limiting page looks like:
rate limiting page
I also took the liberty of refactoring the tabs into a reusable component.

In addition to this, there was [a comment I left on the issue](https://github.com/LemmyNet/lemmy-ui/issues/981#issuecomment-1533974633) that I'm not sure anyone saw:

> I'm wondering if only rate limiting should be in its own tab, or if it should be split up even more. Currently, the site tab on the admin page takes up a lot of vertical space and has a lot of information: this means users have to scan through a bunch of stuff to find the setting they want and, when they want to save, have to scroll all the way to the bottom of a long page. Compartmentalizing the settings into a few more tabs should make it both easier to find the setting you're looking for and reduce the scroll distance needed to reach the save button.
> 
> Here's what I'm thinking for compartmentalizing the Site tab settings into:
> 
> ### Personalization
> - Name
> - Icon
> - Banner
> - Description
> - Theme
> - Sidebar
> - Legal
> 
> ### General
> - Enable Downvotes
> - Enable NSFW
> - Registration Mode
> - Only admins can create communities
> - Require email verification
> - Email admins on receiving new applications
> - Email admins on receiving new reports
> - Private Instance
> - Hide modlog mod names
> - Slur filter regex
> - Languages
> - Max actor name length
> - Captcha enabled
> 
> ### Federation
> - Federation enabled
> - Allowed/blocked instances
> - Federation debug mode
> - Federation worker count
> 
> ### Rate Limiting
> All the rate limiting options.
> 
> ### Taglines
>
> ### Emoji
> Let me know what you think.

Considering those changes would be outside the scope of the rate limit tab, I didn't implement them. Thoughts on the proposed change @dessalines @Nutomic?

### CODE REVIEW

File: src/shared/components/common/tabs.tsx [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+interface TabsProps {
+  tabs: TabItem[];
+}
+
+interface TabsState {
+  currentTab: string;
+}
+
+function handleSwitchTab({ ctx, tab }: { ctx: Tabs; tab: string }) {
+  console.log(tab);
```
Comment (dessalines) [inline src/shared/components/common/tabs.tsx:18]: Remove these console logs after you've got it working.

File: src/shared/components/person/settings.tsx [hunk 2]
```diff
@@ -253,44 +256,26 @@ export class Settings extends Component<any, SettingsState> {
...
+          path={this.context.router.route.match.url}
+          description={this.documentTitle}
+          image={this.state.saveUserSettingsForm.avatar}
+        />
+        <Tabs
+          tabs={[
+            {
+              key: "settings",
+              label: i18n.t("settings"),
+              getNode: this.userSettings,
```
Comment (dessalines) [inline src/shared/components/person/settings.tsx:270]: Nice.

**Review by dessalines — APPROVED** [review]
Once that console logs removed, feel free to merge.

**Comment by makotech222** [comment]
Related, can you try and break out the banned users list into its own tab and make it cleaner? hexbear has thousands and it blows up the page.

**Comment by SleeplessOne1917** [comment]
@makotech222 I can do that as part of that reorganization in a separate PR.

**Comment by SleeplessOne1917** [comment]
@dessalines I just realized I might need to add some translation before I merge. I'm wondering if should take the strings such as "Message rate limit" and use just "Rate Limit" as the label of the top input and use "Member", "Post", etc. as the tab names. Thoughts?

File: src/shared/components/home/rate-limit-form.tsx [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+        search,
+        search_per_second,
+      },
+    };
+  }
+
+  render() {
+    return (
+      <form onSubmit={linkEvent(this, submitRateLimitForm)}>
+        <h5>Rate Limit Options</h5>
```
Comment (dessalines) [inline src/shared/components/home/rate-limit-form.tsx:174]: i18n here too.

**Comment by dessalines** [comment]
> I'm wondering if should take the strings such as "Message rate limit" and use just "Rate Limit" as the label of the top input and use "Member", "Post", etc. as the tab names. Thoughts?

Up to you, whichever is easiest.

**Comment by dessalines** [comment]
K translations are merged now.

**Pull request merged**